File: kernel/gui/render/font.c
#include "font.h"

extern const uint8_t font8x16[256][16];

void font_draw_char(struct surface *s, int x, int y, char c,
                    uint32_t fg, uint32_t bg) {
    const uint8_t *g = font8x16[(uint8_t)c];
    for (int row = 0; row < FONT_H; row++)
        for (int col = 0; col < FONT_W; col++) {
            uint32_t color = (g[row] >> (7 - col)) & 1 ? fg : bg;
            int px = x + col, py = y + row;
            if (px >= 0 && px < s->w && py >= 0 && py < s->h)
                s->pixels[py * s->w + px] = color;
        }
}

void font_draw_string(struct surface *s, int x, int y, const char *str,
                      uint32_t fg, uint32_t bg) {
    while (*str && x + FONT_W <= s->w) {
        font_draw_char(s, x, y, *str++, fg, bg);
        x += FONT_W;
    }
}

File: kernel-rs/src/gui/font.rs
//! 8x16 VGA font drawing onto surfaces.

use super::surface::Surface;
use crate::gui::font8x16_data::FONT8X16;

pub const FONT_W: i32 = 8;
pub const FONT_H: i32 = 16;

pub fn draw_char(s: &mut Surface, x: i32, y: i32, c: u8, fg: u32, bg: u32) {
    let g = &FONT8X16[c as usize];
    for row in 0..FONT_H {
        for col in 0..FONT_W {
            let color = if (g[row as usize] >> (7 - col)) & 1 != 0 {
                fg
            } else {
                bg
            };
            let px = x + col;
            let py = y + row;
            if px >= 0 && px < s.w && py >= 0 && py < s.h {
                unsafe {
                    *s.pixels.add((py * s.w + px) as usize) = color;
                }
            }
        }
    }
}

pub fn draw_string(s: &mut Surface, mut x: i32, y: i32, str_: &[u8], fg: u32, bg: u32) {
    for &c in str_ {
        if c == 0 || x + FONT_W > s.w {
            break;
        }
        draw_char(s, x, y, c, fg, bg);
        x += FONT_W;
    }
}
